prover/src/gpu.rs

```diff
@@ -0,0 +1,284 @@
+where
+    R: RandomCoin<BaseField = Felt, Hasher = Rpo256>,
+{
+    type Air = <ExecutionProver<Rpo256, R> as Prover>::Air;
+    type BaseField = Felt;
+    type Trace = <ExecutionProver<Rpo256, R> as Prover>::Trace;
+    type HashFn = Rpo256;
+    type RandomCoin = R;
+
+    fn options(&self) -> &WinterProofOptions {
+        self.0.options()
+    }
+
+    fn get_pub_inputs(&self, trace: &ExecutionTrace) -> PublicInputs {
+        self.0.get_pub_inputs(trace)
+    }
+
+    /// Computes a low-degree extension (LDE) of the provided execution trace over the specified
+    /// domain and builds a commitment to the extended trace.
+    ///
+    /// The extension is performed by interpolating each column of the execution trace into a
+    /// polynomial of degree = trace_length - 1, and then evaluating the polynomial over the LDE
+    /// domain.
+    ///
+    /// Trace commitment is computed by hashing each row of the extended execution trace, and then
+    /// building a Merkle tree from the resulting hashes.
+    ///
+    /// Interpolations and evaluations are computed on the CPU while hashes are simultaneously
+    /// computed on the GPU:
+    ///
+    /// ```text
+    ///        ──────────────────────────────────────────────────────
+    ///               ┌───┐   ┌────┐   ┌───┐   ┌────┐   ┌───┐
+    ///  CPU:   ... ──┤fft├─┬─┤ifft├───┤fft├─┬─┤ifft├───┤fft├─┬─ ...
+    ///               └───┘ │ └────┘   └───┘ │ └────┘   └───┘ │
+    ///        ╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴
+    ///                     │ ┌──────────┐   │ ┌──────────┐   │
+    ///  GPU:               └─┤   hash   │   └─┤   hash   │   └─ ...
+    ///                       └──────────┘     └──────────┘
+    ///        ────┼────────┼────────┼────────┼────────┼────────┼────
+    ///           t=n     t=n+1    t=n+2     t=n+3   t=n+4    t=n+5
+    /// ```
+    fn build_trace_commitment<E>(
+        &self,
+        trace: &ColMatrix<E>,
+        domain: &StarkDomain<Felt>,
+    ) -> (RowMatrix<E>, MerkleTree<Self::HashFn>, ColMatrix<E>)
+    where
+        E: air::FieldElement<BaseField = Felt>,
+    {
+        // interpolate the execution trace
+        let now = Instant::now();
+        let poly_size = trace.num_rows();
+        let inv_twiddles = fft::get_inv_twiddles::<E::BaseField>(trace.num_rows());
+        let trace_polys = trace.columns().map(|col| {
+            let mut poly = col.to_vec();
+            fft::interpolate_poly(&mut poly, &inv_twiddles);
+            poly
+        });
+
+        // extend the execution trace
+        let lde_segments = FrozenVec::new();
+        let lde_blowup = domain.trace_to_lde_blowup();
+        let offsets = get_evaluation_offsets::<E>(poly_size, lde_blowup, domain.offset());
+        let domain_size = offsets.len();
+        let trace_twiddles = domain.trace_twiddles();
+        let mut lde_segment_generator =
+            SegmentGenerator::new(trace_polys, &offsets, trace_twiddles);
```

Comment:
I think most of the variables declared here are needed only to instantiate `SegmentGenerator`. If this is correct, I'd move most of this code into the constructor of `SegmentGenerator`, and maybe the call here could look like:
```Rust
let mut lde_segment_generator = SegmentGenerator::new(&trace_polys, domain);
```

Also, minor nit: we should probably update the comment above this block as this doesn't actually extend the trace yet.